File: haskell/2025/day07a/src/Main.hs
module Main ( main ) where

import Control.DeepSeq
import Data.Ix
import Data.Set ( Set )
import qualified Data.Set as Set
import System.Clock
import System.Environment
import System.Exit
import System.IO

type Position = (Int, Int)

process :: String -> Int
process content =
    let ls = lines content
        rows = length ls
        cols = length (head ls)
        bounds = ((0, 0), (rows - 1, cols - 1))
        start = head [ (r, c) | r <- [0..rows - 1], c <- [0..cols - 1], ls !! r !! c == 'S' ]
        splitters = Set.fromList [ (r, c) | r <- [0..rows - 1], c <- [0..cols - 1], ls !! r !! c == '^' ]
        currentBeams = Set.singleton start
    in Set.size (simulate bounds splitters currentBeams Set.empty)
  where
    simulate :: ((Int, Int), (Int, Int)) -> Set Position -> Set Position -> Set Position -> Set Position
    simulate bounds splitters currentBeams splitLocs
      | Set.null currentBeams = splitLocs
      | otherwise =
            let (nextBeams, nextSplitLocs) =
                    Set.foldr (\beam@(br, bc) (bs, sls) ->
                                   if beam `Set.member` splitters
                                       then let beamLeft = (br, bc - 1)
                                                beamRight = (br, bc + 1)
                                                bs' = Set.union bs (Set.fromList [beamLeft, beamRight])
                                                sls' = Set.union sls (Set.singleton beam)
                                            in (bs', sls')
                                       else if inRange bounds (br + 1, bc)
                                                then (Set.union bs (Set.singleton (br + 1, bc)), sls)
                                                else (bs, sls))
                              (Set.empty, splitLocs)
                              currentBeams
            in simulate bounds splitters nextBeams nextSplitLocs

usage :: String -> IO ()
usage progname = do
    hPutStrLn stderr $ "usage: " ++ progname ++ " <input file>"
    exitFailure

showTime :: TimeSpec -> String
showTime elapsed =
    let ns = fromIntegral (toNanoSecs elapsed) :: Double
    in if ns < 1000
       then show ns ++ " ns"
       else if ns < 1000000
       then show (ns / 1000.0) ++ " μs"
       else if ns < 1000000000
            then show (ns / 1000000.0) ++ " ms"
            else show (ns / 1000000000.0) ++ " s"

main :: IO ()
main = do
    args <- getArgs
    progname <- getProgName
    case args of
        [filename] -> do
            start <- getTime Monotonic
            content <- readFile filename
            let result = process content
            result `deepseq` return ()
            end <- getTime Monotonic
            let elapsed = diffTimeSpec start end
            putStrLn $ "result = " ++ show result
            putStrLn $ "elapsed time: " ++ showTime elapsed
        _ -> usage progname

File: rust/2025/day07a/src/main.rs
use std::collections::HashSet;
use std::env;
use std::fs;
use std::io;
use std::process;
use std::time::Instant;

type Position = (i64, i64);

fn usage(progname: &str) {
    eprintln!("usage: {progname} <input file>");
    process::exit(1);
}

fn in_range(bounds: &(Position, Position), pos: &Position) -> bool {
    let ((lbr, lbc), (ubr, ubc)) = bounds;
    let (pr, pc) = pos;
    pr >= lbr && pr <= ubr && pc >= lbc && pc <= ubc
}

fn simulate(
    bounds: &(Position, Position),
    splitters: &HashSet<Position>,
    start: Position,
) -> HashSet<Position> {
    let mut stack = Vec::new();
    let mut visited = HashSet::new();
    let mut split_locs = HashSet::new();
    stack.push(start);
    while let Some((r, c)) = stack.pop() {
        let pos = (r, c);
        if !in_range(bounds, &pos) {
            continue;
        }
        if !visited.insert(pos) {
            continue;
        }
        if splitters.contains(&pos) {
            split_locs.insert(pos);
            stack.push((r, c - 1));
            stack.push((r, c + 1));
        } else {
            stack.push((r + 1, c));
        }
    }
    split_locs
}

fn process_file(filename: &str) -> io::Result<i64> {
    let contents = fs::read_to_string(filename)?;
    let rows: i64 = contents.lines().count() as i64;
    let cols: i64 = contents.lines().next().unwrap().chars().count() as i64;
    let bounds = ((0, 0), (rows, cols));
    let mut start: Option<Position> = None;
    let mut splitters: HashSet<Position> = HashSet::new();
    for (row, line) in contents.lines().enumerate() {
        for (col, ch) in line.chars().enumerate() {
            match ch {
                'S' => start = Some((row as i64, col as i64)),
                '^' => {
                    let _ = splitters.insert((row as i64, col as i64));
                }
                _ => {}
            }
        }
    }
    if start.is_none() {
        panic!("no start position found in input!");
    }
    let split_locs = simulate(&bounds, &splitters, start.unwrap());
    Ok(split_locs.len() as i64)
}

fn display_duration(duration: u128) {
    print!("elapsed time: ");
    let duration_f = duration as f64;
    if duration_f < 1_000.0 {
        println!("{} ns", duration_f);
    } else if duration_f < 1_000_000.0 {
        println!("{} μs", duration_f / 1_000.0);
    } else if duration_f < 1_000_000_000.0 {
        println!("{} ms", duration_f / 1_000_000.0);
    } else {
        println!("{} s", duration_f / 1_000_000_000.0);
    }
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let progname = &args[0];
    if args.len() < 2 {
        usage(progname);
    }
    let filename = &args[1];
    let start = Instant::now();
    match process_file(filename) {
        Ok(result) => {
            let duration = start.elapsed().as_nanos();
            println!("result = {result}");
            display_duration(duration);
        }
        Err(e) => {
            eprintln!("error while processing file `{filename}': {e}");
            process::exit(1);
        }
    }
}
